processor/metric.go

```diff
@@ -0,0 +1,56 @@
+package processor
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/elastic/beats/libbeat/monitoring"
+)
+
+type metricMap struct {
+	registry   *monitoring.Registry
+	metricKeys []string
+	mu         sync.Mutex //guards agent creation
+	metrics    map[string]metric
+}
+
+func NewMetricMap(r *monitoring.Registry, metricKeys []string) *metricMap {
+	m := metricMap{registry: r, metricKeys: metricKeys, metrics: map[string]metric{}}
+	return &m
+}
+
+func (m *metricMap) Add(agent, key string, count int) {
+	if val := m.val(agent, key); val != nil {
+		val.Add(int64(count))
+	}
+}
+
+func (m *metricMap) Inc(agent, key string) {
+	if val := m.val(agent, key); val != nil {
+		val.Inc()
+	}
+}
+
+func (m *metricMap) val(agent, key string) *monitoring.Int {
+	r, ok := m.metrics[agent]
+	if !ok {
+		m.mu.Lock()
+		defer m.mu.Unlock()
+		r, ok = m.metrics[agent]
```

Comment:
excellent 👍

Reply:
Sure with the config I can create it upfront.

However,, just for clarification - I don't understand why this should not be sure. 
Line 35 reads from the map, if it doesn't find the key, it uses a `sync.Mutex.Lock` and reads **again** to ensure no other thread is writing while reading. Only if the key is still not find, it is written. The writing always happens within the lock, so there is no scenario where read and write occurs at the same time. Can you elaborate on why this should not be threadsafe?

Reply:
The read on line 35 is not protected by a lock, so while another thread is writing to the map on line 41, a different thread could be reading on line 35